src/hardware/disney.cpp

```diff
@@ -19,456 +20,255 @@
-				disney.da[0].buffer[disney.da[0].used] = disney.data;
-				disney.da[0].used++;
-			} // else LOG_MSG("disney overflow 0");
-		}
-
-		if ((disney.control & 0x8) && !(val & 0x8)) {
-			// emulate a device with 16-byte sound FIFO
-			if (disney.state != DISNEY_STATE::RUNNING) {
-				disney.interface_det_ext++;
-				disney.interface_det = 0;
-				if(disney.interface_det_ext > 5) {
-					disney.leader = &disney.da[0];
-					DISNEY_enable(disney_sample_rate);
-				}
-			}
-			if (disney.interface_det_ext > 5) {
-				if (disney.da[0].used < BUFFER_SAMPLES) {
-					disney.da[0].buffer[disney.da[0].used] = disney.data;
-					disney.da[0].used++;
-				}
-			}
-		}
-
-//		LOG_WARN("DISNEY:Control write %x",val);
-		if (val&0x10) LOG(LOG_MISC,LOG_ERROR)("DISNEY:Parallel IRQ Enabled");
-		disney.control=val;
-		break;
-	}
+	// The Disney ACK's (active-low) when the FIFO has room
+	status_reg.ack = IsFifoFull();
+	return status_reg.data;
 }
 
-static uint8_t disney_read(io_port_t port, io_width_t)
+void Disney::AudioCallback(const uint16_t requested_frames)
 {
-	uint8_t retval;
-	switch (port - DISNEY_BASE) {
-	case 0: /* Data Port */
-		// LOG(LOG_MISC,LOG_NORMAL)("DISNEY:Read from data port");
-		return disney.data;
-		break;
-	case 1:		/* Status Port */
-//		LOG(LOG_MISC,"DISNEY:Read from status port %X",disney.status);
-		retval = 0x07;//0x40; // Stereo-on-1 and (or) New-Stereo DACs present
-		if (disney.interface_det_ext > 5) {
-			if (disney.leader && disney.leader->used >= 16){
-				retval |= 0x40; // ack
-				retval &= ~0x4; // interrupt
-			}
-		}
-		if (!(disney.data&0x80)) retval |= 0x80; // pin 9 is wired to pin 11
-		return retval;
-		break;
-	case 2:		/* Control Port */
-		LOG(LOG_MISC,LOG_NORMAL)("DISNEY:Read from control port");
-		return disney.control;
-		break;
+	assert(channel);
+
+	auto frames_remaining = requested_frames;
+
+	// First, add any frames we've queued since the last callback
+	while (frames_remaining && render_queue.size()) {
+		channel->AddSamples_sfloat(1, &render_queue.front()[0]);
+		render_queue.pop();
+		--frames_remaining;
+	}
+	// If the queue's run dry, render the remainder and sync-up our time datum
+	while (frames_remaining) {
+		const auto frame = Render();
+		channel->AddSamples_sfloat(1, &frame[0]);
+		--frames_remaining;
 	}
-	return 0xff;
+	last_rendered_ms = PIC_FullIndex();
 }
 
-static void DISNEY_PlayStereo(uint16_t len, const uint8_t *l, const uint8_t *r)
+std::unique_ptr<Disney> disney = {};
+
+static void setup_filters(mixer_channel_t &channel)
 {
-	static uint8_t stereodata[BUFFER_SAMPLES * 2];
-	for (uint16_t i = 0; i < len; ++i) {
-		stereodata[i*2] = l[i];
-		stereodata[i*2+1] = r[i];
-	}
-	disney.chan->AddSamples_s8(len,stereodata);
+	// The filters are meant to emulate the Disney's bandwidth limitations
+	// both by ear and spectrum analysis when compared against LGR Oddware's
+	// recordings of an authentic Disney Sound Source in ref:
+	// https://youtu.be/A1YThKmV2dk?t=1126
```

Comment:
Great comment! I'm contemplating of adding similar links to my older filter related comments.

Reply:
Thanks! It did feel a bit strange.. dropping a youtube link into a C++ comments :-) But hey, it's what was used, here and now, for the purists out there who might want to check these things (and for historical record). All interesting facts!